src/java/org/apache/cassandra/cql3/statements/schema/CreateIndexStatement.java

```diff
@@ -99,6 +112,18 @@ public Keyspaces apply(Keyspaces schema)
         if (Keyspace.open(table.keyspace).getReplicationStrategy().hasTransientReplicas())
             throw new InvalidRequestException("Secondary indexes are not supported on transiently replicated keyspaces");
 
+        // guardrails to limit number of secondary indexes per table.
+        if (!attrs.isCustom)
+        {
+            long existingSecondaryIndexes = table.indexes.stream().filter(indexMetadata -> !indexMetadata.isCustom()).count();
+            Guardrails.secondaryIndexesPerTable.guard(existingSecondaryIndexes + 1,
+                                                      Strings.isNullOrEmpty(indexName)
+                                                      ? String.format("on table %s", table.name)
+                                                      : String.format("%s on table %s", indexName, table.name),
+                                                      false,
+                                                      state);
+        }
```

Comment:
SASI is in-tree and SAI is planned to also be in-tree; do you see each implementation requiring their own guardrails?  What about users who mix-and-match implementations; should we not have a global limit?

Reply:
Indeed I think that each implementation might require its own guardrail, but we are not there yet. I have modified this guardrail to also include custom indexes, so we can use it as a global limit for the number of indexes of any kind. Later we can decide if we want to add specific guardrails for SAI, SASI or legacy indexes.